app/components/dataurl-display.tsx

```diff
@@ -0,0 +1,101 @@
+import { useCallback, useRef, useState } from "react";
+import Button from "~/components/button";
+import { Transition } from "@headlessui/react";
+import Copy from "~/components/copy";
+
+interface Props {
+  readonly label: string;
+  readonly display: (text: string) => Promise<string>;
+  readonly rows: number;
+}
+
+export default function DataUrlDisplay({ label, display, rows }: Props) {
+  const [output, setOutput] = useState<string | null>(null);
+  const [error, setError] = useState<string | null>(null);
+  const inputRef = useRef<HTMLTextAreaElement>(null);
+
+  const process = useCallback(() => {
+    const text = inputRef.current?.value || "";
+
+    display(text)
+      .then((it) => {
+        if (it.length > 0) {
+          setOutput(it);
+          setError(null);
+        } else {
+          setError("Does not appear to be a valid data URL.");
+          setOutput(null);
+        }
+      })
+      .catch((e) => {
+        setError(e.message);
+        setOutput(null);
+      });
+  }, [display]);
```

Comment:
I think we might need to handle this better. I know this is somewhat copied from the code I have, but there has to be a better way here.

The best way tto handle errors in React is error boundaries, maybe that could work here if we just had one component that was responsible of taking the input string and drawing the output display. It could either draw the output, or the error boundary could show the error.

That might be too large a change, but the easy option might be to have the `display` function be responsible for not returning an empty string, and have it throw an error if the string is empty which should be caught by the `catch` function that is already there.